src/uschedulecmd.1.in
.TH uschedulecmd 1 \*(VE \*(PA      
.hy 0
.URL "Uwe Ohse" ../uwe.html
.br
.URL "The uschedule package" ../uschedule.html
.paragraph
.SH NAME
uschedulecmd - create a new uschedule jobs
.LINE
.section SYNOPSIS
\fBuschedulecmd\fR [-.] [-d DIR] [-e] [-i ID] [COMMAND ...]
.section DESCRIPTION
\fBuschedulecmd\fR creates a new job. If \fICOMMAND\fR is not given
then command to be executed will be read from the standard input.
.section OPTIONS
.options
.option "-., --dot-as-home"
The current working directory will be used instead of $HOME.
.option "-d, --dir=DIR"
Put the new job into \fIDIR\fR. The default is \fI~/.uschedule/\fR.
.option "-e, --copy-environment"
Clone the current environment for the new job.
.option "-i, --id=ID"
Give the new job the identifier \fIID\dR. \fIID\fR must be unique
and must not contain colons, slashes and dots.
.br
If this option is not used then a unique numerical ID will be given
and printed to the standard output.
.section CAVEAT
Note that this just creates the command to be run, it doesn't 
schedule the job to be run at any time. Use the
.manlink uschedule.html uschedule(1)
command for this.
.section AUTHOR
Uwe Ohse, uwe@ohse.de
.section "SEE ALSO"
.manlink uschedule.html uschedule(1) ,
.manlink intro.html uschedule_intro(7) .
.HTML <!--
.paragraph
The homepage may be more up-to-date, see
.br
http://www.ohse.de/uwe/uschedule.html.
.HTML -->


src/NEWS
Version 0.7.1, August 2004:

* fixed uschedulelist --short (-s), which used to display the "next" 
  instead of the "last" time.
* "uschedule --every X" (-e) repeats a job with a given interval X.
* The new uscheduleruntimelimit tool allows to limit the run time of a job.
* uscheduleconf makes base directory sticky for daemontools <0.75.
* uschedulelist --show-command (-S) shows the command in addition to the 
  usual list output.
* uschedule accepts + as synonym for / in TIMESPEC.

Version 0.7.0, March 2003:

* implemented -. option ("use . instead of $HOME").
* bugfix, "-" as username or logusername was mistreated.

  *** Impact: the user may have got root rights. ***

  Note: older versions may have left a multilog running as root instead
  of the user. Impact: the user may have got root rights.
  Fix: Go though /service/*uschedule*/log/run and add 
    setuidgid "username" \
  after the line containing the softlimit call and before the line 
  containing run. Afterwards the file should look like this:
    #! /bin/sh
    exec 2>&1
    cd '...SOMEWHERE...' || exit 1
    exec \
    softlimit -m 8000000 -o 400 -p 40 \
    setuidgid USERNAME \
    ./run

  The code containing the bug has been rewritten and simplified.

* uscheduleconf now has a --no-user-change option.

Version 0.6.1, October 2002:

* installed manual pages again.

Version 0.6.0, October 2002:

* uschedulecmd used to steal and misinterpret options given to the child 
  command. Now all programs stop parsing looking for options after the
  first non-option argument.
* implemented start/end times for jobs. See the uschedule manual page
  for more information.
* uschedulelist now prints all month in the 1..12 range, not 0..11 anymore.

Version 0.5.8, November 2001:

* internal changes to get rid of some warnings.
* the same format is used whereever a time is printed.

Version 0.5.7, November 2001:

* uscheduleconf created completely broken configuration if a log account
  different from the user account was given.

Version 0.5.6, October 2001:

* uscheduleconf created buggy call to multilog.

Version 0.5.5, October 2001:

* fixed another self check problem.

Version 0.5.4, October 2001:

* fixed a number of documentation bugs.
* uscheduleconf now starts uscheduled instead of the nonexistant scheduled.
* self check now works if the machine has a time zone different from UTC
  or the local file system has a block size different from 8k.

Version 0.5.3, September 2001:

* first public release.

Version 0.5.1 to 0.5.2, September 2001:

* internal release.

Version 0.5.0, August 2001:

* first internal release.


src/uschedulecmd.c
#include "uogetopt.h"
#include "scheduled.h"
#include "open.h"
#include "attributes.h"
#include "stralloc.h"
#include "bailout.h"
#include "buffer.h"
#include "env.h"
#include "str.h"
#include "scan.h"
#include "error.h"
#include <unistd.h> /* read, write, fsync, chdir */
#include <sys/stat.h> /* fchmod */

static const char *o_id;
static const char *o_dir;
static int o_dot_as_home;
static int o_env=0;

static uogetopt2 myopts[]={
{'.',"dot-as-home",uogo_flag,UOGO_NOARG,&o_dot_as_home, 1 ,
"Use current directory instead of $HOME.",0,0},
{'d',"dir",uogo_string,0,&o_dir, 0 ,
"Jobdirectory.",
"Jobs will be read from this directory.","DIR"},
{'e',"copy-environment",uogo_flag,UOGO_NOARG,&o_env, 1 ,
"Clone current environment for job.",
"The default is to start new jobs in the daemons environment. This option "
"causes the current environment to be copied for the new process.",0},
{'i',"id",uogo_string,0,&o_id, 0 ,
"Unique ID for the new job.",
"An identifier must not be re-used and must not contain colons, slashes and "
"dots. If this option is not used then a unique ID will be given.", "ID"},
{0,0,0,0,0,0,0,0,0}
};              
static void die_write(void) attribute_noreturn;
static void die_write_temp(void) attribute_noreturn;
static void die_fsync(void) attribute_noreturn;
static void die_fchmod(void) attribute_noreturn;

static void die_write(void)  {xbailout(111,errno,"failed to write",0,0,0);}
static void die_write_temp(void)
{ xbailout(111,errno,"failed to write temporary file",0,0,0); }
static void die_fsync(void)
{ xbailout(111,errno,"failed to fsync temporary file",0,0,0); }
static void die_fchmod(void)
{ xbailout(111,errno,"failed to fchmod temporary file",0,0,0); }

extern char **environ;
#define W(b,s,l) do { \
 if (-1==buffer_put(b,s,l)) die_write_temp(); \
} while (0)

static void
copy_env(buffer *o)
{
	unsigned int i;
	if (-1==buffer_puts(o,"env - \\\n")) die_write_temp();
	for (i=0;environ[i];i++) {
		char *e=environ[i];
		unsigned int j;
		j=str_chr(e,'=');
		if (!e[j])
			continue;
		W(o,e,j+1);
		W(o,"'",1);
		j++;
		while (e[j]) {
			if (e[j]=='\'')
				W(o,"'\\'",3);
			W(o,e+j,1);
			j++;
		}
		W(o,"' \\\n",4);
	}
}


static void
create_job(int fd, char **args)
{
	buffer i;
	char ispace[512];
	buffer o;
	char ospace[512];

	buffer_init(&o,(buffer_op) write,fd,ospace,sizeof(ospace)); 
	if (-1==buffer_puts(&o,"#! /bin/sh\n")) die_write_temp();
	if (o_env) {
		static stralloc fn;
		if (!stralloc_copys(&fn,o_id)) oom();
		if (!stralloc_cats(&fn,".run")) oom();
		if (!stralloc_0(&fn)) oom();

		if (-1==buffer_puts(&o,"exec \\\n")) die_write_temp();
		copy_env(&o);
		if (-1==buffer_puts(&o,"/bin/sh -c ")) die_write_temp();
		if (-1==buffer_puts(&o,"\"$SCHEDULED_COMMAND_FILE\"/'")) 
			die_write_temp();
		if (-1==buffer_puts(&o,fn.s)) die_write_temp();
		if (-1==buffer_puts(&o,"'\n")) die_write_temp();
		if (-1==buffer_flush(&o)) die_write_temp();
		if (-1==fchmod(fd,0755)) die_fchmod();
		if (-1==fsync(fd)) die_fsync();
		if (-1==close(fd)) die_write_temp();
		fd=open_excl(fn.s);
		if (fd==-1)
			xbailout(111,errno,"failed to open_excl ",fn.s,0,0);
		buffer_init(&o,(buffer_op) write,fd,ospace,sizeof(ospace)); 
		if (-1==buffer_puts(&o,"#! /bin/sh\n")) die_write_temp();
	} 

	if (args) {
		unsigned int j;
		const char *spc="";
		for (j=0;args[j];j++) {
			if (-1==buffer_puts(&o,spc)) die_write_temp();
			if (-1==buffer_puts(&o,args[j])) die_write_temp();
			spc=" ";
		}
	} else {
		buffer_init(&i,(buffer_op) read,0,ispace,sizeof(ispace)); 
		switch (buffer_copy(&o,&i)) {
		case 0: break;
		case -2: xbailout(111,errno,"failed to read input",0,0,0);
		case -3: die_write_temp();
		default: xbailout(100,errno,"unexpected errorcode from buffer_copy",0,0,0);
		}
	}
	if (-1==buffer_puts(&o,"\n")) die_write_temp();
	if (-1==buffer_flush(&o)) die_write_temp();
	if (-1==fchmod(fd,0755)) die_fchmod();
	if (-1==fsync(fd)) die_fsync();
	if (-1==close(fd)) die_write_temp();
}
static uogetopt_env myoptenv={
"uschedulecmd",PACKAGE,VERSION,
"uschedulecmd [options] [COMMAND ...]",
"This program creates a schedule command executing COMMAND or, if that is "
"not given, a command read from the standard input.",
"long",
"Report bugs to uschedule@lists.ohse.de",
0,0,0,0,uogetopt_out,myopts
};


int main(int argc, char **argv) 
{
	int fd=-1; /* keep gcc quiet */
	int prt=0;

	bailout_progname(argv[0]);
	flag_bailout_fatal_begin=3;
	uogo_posixmode=1;
	myoptenv.program=flag_bailout_log_name;
	uogetopt_parse(&myoptenv,&argc,argv);

	change_dir(0,o_dir,o_dot_as_home);
	if (-1==chdir(IDDIR)) 
		xbailout(111,errno,"failed to chdir to ",IDDIR, " directory",0);

	if (!o_id) {
		static stralloc sat;
		fd=make_id(&sat);
		o_id=sat.s;
		prt=1;
	} else {
		check_id(o_id);
		fd=open_excl(o_id);
		if (-1==fd)
			xbailout(111,errno,"failed to open_excl ",o_id,0,0);
	}

	create_job(fd,argv[1] ? argv+1 : 0);
	if (prt) {
		if (-1==buffer_puts(buffer_1,"The id of the new job is ")) die_write();
		if (-1==buffer_puts(buffer_1,o_id)) die_write();
		if (-1==buffer_puts(buffer_1,"\n")) die_write();      
		if (-1==buffer_flush(buffer_1)) die_write();      
	}
	return 0;
}
